src/Features/Core/Portable/FindUsages/AbstractFindUsagesService.ProgressAdapter.cs

```diff
@@ -107,20 +108,26 @@ public async ValueTask OnDefinitionFoundAsync(SymbolGroup group, CancellationTok
         }
 
         public async ValueTask OnReferencesFoundAsync(
-            ImmutableArray<(SymbolGroup group, ISymbol symbol, ReferenceLocation location)> references, CancellationToken cancellationToken)
+            IAsyncEnumerable<(SymbolGroup group, ISymbol symbol, ReferenceLocation location)> references, CancellationToken cancellationToken)
         {
-            foreach (var (group, _, location) in references)
-            {
-                var definitionItem = await GetDefinitionItemAsync(group, cancellationToken).ConfigureAwait(false);
-                var referenceItem = await location.TryCreateSourceReferenceItemAsync(
-                    classificationOptions,
-                    definitionItem,
-                    includeHiddenLocations: false,
-                    cancellationToken).ConfigureAwait(false);
-
-                if (referenceItem != null)
-                    await context.OnReferenceFoundAsync(referenceItem, cancellationToken).ConfigureAwait(false);
-            }
+            await ProducerConsumer<SourceReferenceItem>.RunParallelAsync(
+                source: references,
+                produceItems: static async (tuple, callback, args, cancellationToken) =>
+                {
+                    var (group, _, location) = tuple;
+                    var definitionItem = await args.@this.GetDefinitionItemAsync(group, cancellationToken).ConfigureAwait(false);
+                    var sourceReferenceItem = await location.TryCreateSourceReferenceItemAsync(
+                        args.classificationOptions,
+                        definitionItem,
+                        includeHiddenLocations: false,
+                        cancellationToken).ConfigureAwait(false);
+                    if (sourceReferenceItem != null)
+                        callback(sourceReferenceItem);
+                },
+                consumeItems: static async (items, args, cancellationToken) =>
+                    await args.context.OnReferencesFoundAsync(items, cancellationToken).ConfigureAwait(false),
+                args: (@this: this, context, classificationOptions),
```

Comment:
>context, classificationOptions [](http://example.com/codeflow?start=36&length=30)

nit: Looks like both of these are defined at the class level and you are already passing in this.

Reply:
these are primary constructor params.  they are not available off of `this`.

Reply:
I still find primary ctors and the names we use because of them quite unintuitive. I'll eventually acclimate, but not today I guess.